### PR TITLE
Agent infrastructure

### PR DESCRIPTION
**Pull request by mvidner**
https://trello.com/c/MMlaM6sK/6-better-agent-infrastructure

### CODE REVIEW

File: library/general/src/lib/config_file/augeas_parser.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+require "augeas"
+
+module ConfigAgent
+  class AugeasParser
```
Comment (mvidner) [inline library/general/src/lib/config_file/augeas_parser.rb]: > \# TODO: figure out a way to bind to Augeas, and to replace ConfigFile::KeyValueParser with it

File: library/general/src/lib/config_file/augeas_parser.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require "augeas"
+
+module ConfigAgent
```
Comment (mvidner) [inline library/general/src/lib/config_file/augeas_parser.rb]: s/ConfigAgent/ConfigFile/ ?

File: library/general/src/lib/config_file/config_file.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+module ConfigFile
+  class ConfigFile
```
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: About naming:
- `ConfigFile` for the whole module sounds good.
- `ConfigFile::ConfigFile` -> `ConfigFile::ModelBase`
- `GrubConfig` -> `GrubModel`
- `KeyValue_Parser` -> `KeyValueParser`
- `AugeasParser` sounds good

File: library/general/src/lib/config_file/grub_config.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
+require "config_file/config_file"
+require "config_file/keyvalue_parser"
+
+module ConfigFile
+  PARSER = Keyvalue_Parser
+  PATH = "/etc/default/grub"
+
+  class GrubConfig < ConfigFile
```
Comment (mvidner) [inline library/general/src/lib/config_file/grub_config.rb]: `GrubModel`

File: library/general/src/lib/config_file/keyvalue_parser.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+module ConfigFile
+  class Keyvalue_Parser
```
Comment (mvidner) [inline library/general/src/lib/config_file/keyvalue_parser.rb]: It's a Camelo_Snake! `KeyValueParser`

File: library/general/src/lib/config_file/config_file.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    def initialize(parser, file_path, file_class: File)
+      @file_class = file_class
+      @parser = parser
+      @file_path = file_path
+    end
+
+    def write(changes_only: false)
```
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: So `read`, `write` are part of the public API... Wouldn't it be better to adopt an ActiveModel API, with `save` instead?

File: library/general/src/lib/config_file/config_file.rb (hunk 3 above)
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: > \# A base class for models. Represents a configuration file as an object with domain-specific attributes/methods.
Uses
- a **file** (usually plain `::File`) to convert between the on-disk storage and a `String`,
- a **parser** (usually specific for each derived Model) to convert between that `String` and a private representation

File: library/general/src/lib/config_file/keyvalue_parser.rb (hunk 5 above)
Comment (mvidner) [inline library/general/src/lib/config_file/keyvalue_parser.rb]: > \# A `ConfigFile::ParserBase` can `parse` and `serialize` some data between a String and some structured representation of that string. According to the type of that representation it is coupled with a class derived from `ConfigFile::ModelBase`

File: library/general/src/lib/config_file/config_file.rb (hunk 6 above)
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: What is our position on caching and flushing? (compare the explicit flushes required with some SCR agents, (not) re-reading the config on repeated Read)

File: library/general/src/lib/config_file/augeas_parser.rb (hunk 2 above)
Comment (jreidinger) [inline library/general/src/lib/config_file/augeas_parser.rb]: yeah, typo, original name and then I recognize I do not like agent term, when it is in fact configuration file.

File: library/general/src/lib/config_file/grub_config.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      data["GRUB_CMDLINE_LINUX_RECOVERY"][:commented_out] = true
+    end
+
+    def enable_recovery_entry(kernel_params)
+      data["GRUB_DISABLE_RECOVERY"] ||= {}
+      data["GRUB_DISABLE_RECOVERY"][:value] = "true"
+      data["GRUB_DISABLE_RECOVERY"][:commented_out] = false
+      data["GRUB_CMDLINE_LINUX_RECOVERY"] ||= {}
+      data["GRUB_CMDLINE_LINUX_RECOVERY"][:value] = kernel_params
+      data["GRUB_CMDLINE_LINUX_RECOVERY"][:commented_out] = false
```
Comment (mvidner) [inline library/general/src/lib/config_file/grub_config.rb]: For the record, this part of the model is repetitive and would get refactored for production use, but it isn't our focus now.

File: library/general/src/lib/config_file/config_file.rb (hunk 3 above)
Comment (jreidinger) [inline library/general/src/lib/config_file/config_file.rb]: I do not like `GrubModel` as it sounds like model for whole grub, which is not true...so maybe `GrubConfModel`?

File: library/general/src/lib/config_file/config_file.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+      merge_changes if changes_only
```
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: The merging feature sounds... interesting and complicated. Too early to say more.

File: library/general/src/lib/config_file/keyvalue_parser.rb (hunk 5 above)
Comment (jreidinger) [inline library/general/src/lib/config_file/keyvalue_parser.rb]: yeah, creating name is the biggest challenge :wink:

File: library/general/src/lib/config_file/config_file.rb (hunk 6 above)
Comment (jreidinger) [inline library/general/src/lib/config_file/config_file.rb]: whatever agent needs. In general common base support read/write, no caching, only parsed tree is stored in memory, so you can call more writes...if you need you can call more reads to reset.

**Comment by mvidner** [comment]
Overall it looks good. I think we should work on the external interface before wrestling with the augeas internals: see how it actually replaces existing SCR code, for more than one model.

File: library/general/src/lib/config_file/config_file.rb (hunk 3 above)
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: In general, a matching FooModel and FooParser. Now, does `GrubConfModel` describe /etc/grub.conf? `EtcDefaultGrubModel`? But that's not important for the general API.

Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: > \# A base class for models. Represents a configuration file as an object with domain-specific attributes/methods. For persistent storage, use `load` and `save`,
Non-responsibilities: actual storage and parsing (both delegated).

File: library/general/src/lib/config_file/config_file.rb (hunk 6 above)
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: \# No extra caching (unlike some SCR agents)

File: library/general/src/lib/config_file/config_file.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+      @file_class.write(@file_path, @parser.serialize(data))
+    end
+
+    def read
```
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: \# No extra caching (unlike some SCR agents)

File: library/general/src/lib/config_file/config_file.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+  protected
+
+    attr_accessor :data
+
+    def merge_changes
+      new_data = data.dup
+      read
+      # TODO recursive merge
+      date.merge(new_data)
```
Comment (mvidner) [inline library/general/src/lib/config_file/config_file.rb]: data

**Comment by jreidinger** [comment]
testing script:
```
jreidinger@linux:~/yast/yast2/library/general/src/lib/config_file> cat ~/test.rb 
$: << "/home/jreidinger/yast/yast2/library/general/src/lib"

require "config_file/grub_model"
require "config_file/memory_file"

config = ConfigFile::GrubModel.new
config.load

puts"config: " + config.inspect
puts ""
puts "os prober:  #{config.os_prober_enabled?}"

memory_file = ConfigFile::MemoryFile.new
config2 = ConfigFile::GrubModel.new(file_class:  memory_file)

config2.disable_os_prober
config2.enable_recovery_entry "\"kernel_do_your_job=HARD!\""

config2.save

puts
puts "Testing output:"
puts memory_file.content

```

File: library/general/src/lib/config_file/memory_file.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
+module ConfigFile
+  class MemoryFile
+    attr_accessor :content
+
+    def initialize(content = "")
+      @content = content
+    end
+
+    def read(path)
```
Comment (mvidner) [inline library/general/src/lib/config_file/memory_file.rb]: This being an instance method means you have to pass an instance to a parameter named "class": 
```ruby
memory_file = ConfigFile::MemoryFile.new
config2 = ConfigFile::GrubModel.new(file_class:  memory_file)
```

That's not nice, but I do not see a better option :(

Comment (jreidinger) [inline library/general/src/lib/config_file/memory_file.rb]: maybe just naming issue. I plan to use it this way from beginning.

**Comment by mvidner** [comment]
TODO: try and describe how we handle (un)escaping of values.

**Comment by mvidner** [comment]
OK, so value escaping is probably not necessary with respect to Augeas core, and depends on each individual lens.
But Augeas core gets into play for keys: what if a key contains a slash or a space or a bracket. See https://bugzilla.suse.com/show_bug.cgi?id=925225

**Comment by jreidinger** [comment]
lets close it as now it lives in separate gem at https://github.com/jreidinger/config_files

**Pull request closed without merging**